docs/RFCS/20191009_gc_protected_timestamps.md

```diff
@@ -0,0 +1,597 @@
+timestamps.
+
+This meta row will help us implement this feature by enabling caching and size
+limits. It will be observed and modified in all transactions which create
+protected timestamps. We'll also observe this meta row in all transactions
+which update zone configurations.
+
+Now let's walk through the high level protocol for writing protected timestamps
+and for GC.
+
+#### Protecting a timestamp
+
+First the code will, in the transaction, read all of the relevant zone
+configurations for the spans being protected (this is not as cheap as it should
+be, see TODO other RFC). It will then determine the relevant zone config with
+the shortest TTL. The deadline for the transaction to be committed is the
+requested timestamp plus that minimum TTL.
+
+Every time a protected timestamp is written it must check that it will not
+exceed the rows or spans limits. If the limits would be exceeded an errors is
+returned (see limits below). If the limits are not exceeded, the counters in
+the meta row are incremented the counts accordingly, along with the version,
+then insert the row and spans are inserted.
+
+#### GC
+
+GC occurs on a `Replica` when a heuristic determines that there is
+enough garbage which should be collected. It is important to note that this
+heuristic is guided by the TTL (which sets a lower bound) but is often quite
+a bit later than the TTL. When it is determined that a replica should 
+run GC today then it generally just does it at the latest timestamp allowed
+by the TTL. This RFC is going to require more careful selection of a the 
+GC timestamp.
+
+Given the above described protocol for writing protected timestamp state,
+let's lay out the conditions under which a Replica can GC data.
+
+When running GC the replica will have the following information:
+
+  * Protected TS state
+    - This is exposed to the `Replica` through an interface which exposes:
+    1) MVCC timestamp at which this state was retrieved
+    2) A method to query the set of timestamps which intersect the `Replica`'s 
+       keyspace and their associated `ProtectionMode`
+  * Zone Config
+    - We assume that each range has exactly one zone config.
+    - Sometimes a range will need to be split because it spans more than
+      one zone config but we won't (don't) run GC in that case.
+    - We also have with the zone config the MVCC timestamp at which the
+      zone config was written.
+  * The current time
+
+1) Ensure that the zone config's MVCC timestamp is older than the read time of
+   the Protected TS state, if it's not then no garbage collection can occur
+   until the protected TS state has been refreshed.
+
+1) Read the protected TS state to determine which timestamps are protected
+   for this range. Any read timestamps must be protected regardless of the
+   rest of the procedure.
+
+At this point it is safe to garbage collect up to now less the GC TTL less
+the time since the protected state was read (and a clock max offset).
```

Comment:
Do you mind writing this correctness proof out a bit more formally (no need to go overboard)? Mostly I'm just interested in why a max offset slipped in here and how the MVCC timestamp of the `protected_ts_timestamps` row relates to its `ts` field.